### PR TITLE
Migrate Dashboard demo components to TypeScript

### PR DESCRIPTION
**Pull request by developerium**
Hi guys

I've just refactored a couple of components to typescript. Please check it out and let me know if I'm moving in the right direction.

If I guess correctly, all reusable component types should go inside `examples/demo/src/types.ts` right?

### CODE REVIEW

**Comment by fzaninotto** [comment]
Looking good, and yes, reusable types should go to a central `types.ts` file, just like in ra-core. 

Could you add a checklist to your issue description, so that we know which components you intend to merge, and when the PR is ready to be merged?

**Comment by developerium** [comment]
I'm not sure how to make a checklist, but I'm want to refactor all components in `examples/demo/src/dashboard`

**Comment by fzaninotto** [comment]
Then please add a comment when you think you're done.

**Comment by developerium** [comment]
All done!
Please check it out

File: examples/demo/src/dashboard/Dashboard.tsx [hunk 1]
```diff
@@ -9,6 +15,28 @@ import PendingOrders from './PendingOrders';
...
+    nbNewOrders: number;
+    pendingOrders: Order[];
+}
+
+interface CustomerData {
+    [key: string]: Customer;
+}
+
+interface State {
+    nbNewOrders: number;
```
Comment (fzaninotto) [inline examples/demo/src/dashboard/Dashboard.tsx]: These must be optional, otherwise the initial state (`{}`) won't compile.

Comment (developerium) [reply examples/demo/src/dashboard/Dashboard.tsx]: Done!

Comment (developerium) [reply examples/demo/src/dashboard/Dashboard.tsx]: Interestingly I didn't get any warnings about them in development, I used `make run-demo` command to run it locally

Comment (fzaninotto) [reply examples/demo/src/dashboard/Dashboard.tsx]: TypeScript is overrated :man_shrugging:

**Comment by fzaninotto** [comment]
Thanks!

**Comment by developerium** [comment]
No problem 😄

**Pull request merged**